### ISSUE TITLE
Prediction on batch is slow

### ISSUE THREAD

**Issue by netw0rkf10w** [fizyr/keras-retinanet#537]
Hi,

I have realized that prediction on batch is very slow, just like predicting the images one by one.

Results when passing the images one by one:

> image 0
> 	 reading img.:		 0.30931925773620605
> 	 pre-process:		 0.1686263084411621
> 	 prediction:		 28.559466123580933
> 	 box refining:		 0.017919540405273438
> 	 drawing time:		 0.01849508285522461
> 	 Total time:		 29.07526707649231
> image 1
> 	 reading img.:		 0.6545109748840332
> 	 pre-process:		 0.16336822509765625
> 	 prediction:		 0.4428994655609131
> 	 box refining:		 0.01468205451965332
> 	 drawing time:		 0.016250133514404297
> 	 Total time:		 1.2931923866271973
> image 2
> 	 reading img.:		 0.6143767833709717
> 	 pre-process:		 0.15154314041137695
> 	 prediction:		 0.4159367084503174
> 	 box refining:		 0.00628972053527832
> 	 drawing time:		 0.010651350021362305
> 	 Total time:		 1.1995365619659424
> etc.

Look at the **prediction** line: for each image (except the very first one) it took ~0.4s.

Now for passing images in batches.

For **batch_size = 8**:
> batch 0
> 	 stack:     	 0.07756924629211426
> 	 prediction:	 22.865782737731934
> 	 box refine:	 0.21228313446044922
> 	 draw:      	 0.15601873397827148
> 	 write:     	 1.4322752952575684
> batch 1
> 	 stack:     	 0.058986663818359375
> 	 prediction:	 2.878574848175049
> 	 box refine:	 0.3401339054107666
> 	 draw:      	 0.19519901275634766
> 	 write:     	 1.593874216079712
> batch 2
> 	 stack:     	 0.05459475517272949
> 	 prediction:	 2.8716516494750977
> 	 box refine:	 0.263216495513916
> 	 draw:      	 0.15625524520874023
> 	 write:     	 1.5663535594940186
> etc.

For **batch_size = 16**:
> batch 0
> 	 stack:     	 0.10399198532104492
> 	 prediction:	 15.486302614212036
> 	 box refine:	 0.6394941806793213
> 	 draw:      	 0.35182785987854004
> 	 write:     	 3.0538930892944336
> batch 1
> 	 stack:     	 0.08749055862426758
> 	 prediction:	 4.982178688049316
> 	 box refine:	 0.5139613151550293
> 	 draw:      	 0.34847402572631836
> 	 write:     	 3.0445525646209717
> batch 2
> 	 stack:     	 0.1171884536743164
> 	 prediction:	 4.816910743713379
> 	 box refine:	 0.4572336673736572
> 	 draw:      	 0.32962489128112793
> 	 write:     	 2.9930830001831055
> etc.

For **batch_size = 24**:
> batch 0
> 	 stack:     	 0.1482241153717041
> 	 prediction:	 24.142788648605347
> 	 box refine:	 0.8734383583068848
> 	 draw:      	 0.5006444454193115
> 	 write:     	 4.255364418029785
> batch 1
> 	 stack:     	 0.20960402488708496
> 	 prediction:	 7.7279016971588135
> 	 box refine:	 0.7603216171264648
> 	 draw:      	 0.5628409385681152
> 	 write:     	 4.494097948074341
> batch 2
> 	 stack:     	 0.1456892490386963
> 	 prediction:	 7.624686241149902
> 	 box refine:	 0.5359351634979248
> 	 draw:      	 0.4038214683532715
> 	 write:     	 4.534356355667114

In summary:

> batch_size=1: ~0.4s
> batch_size=8: ~2.9s
> batch_size=16: ~5.0s
> batch_size=24: ~7.7s

The prediction time increases linearly w.r.t. the batch size.

To be clear, the prediction time shown as above is measured for the `predict_on_batch` step only, i.e.:

```
        # prediction
        start = time.time()
        all_boxes, all_scores, all_labels = model.predict_on_batch(image_batch)
        time_prediction = (time.time() - start)
```

I am not sure if this is rather a Keras problem.

Could you please check? Thank you very much in advance!

**Comment by hgaiser** [comment]
Hmm I'm not sure what else you were expecting? Should it be that much faster if you do batch-wise prediction?

From batch_size=1 to batch_size=24 the time increased by 19.25, but the data increased by 24. To me that seems as expected.

What is the reason you think it should be faster?

**Comment by netw0rkf10w** [comment]
@hgaiser I will check again on a classification task in PyTorch and TensorFlow to see how it scales. Need to revise my knowledge :(

**Comment by kbrajwani** [comment]
i am new in this batch process can you provide code how to create batch of image from folder and pass to predict_on_batch
